src/GraphBLAS-sharp.Backend/CSRMatrix/Mxm/Expansion.fs

```diff
@@ -0,0 +1,416 @@
+namespace GraphBLAS.FSharp.Backend
+
+open Brahma.FSharp.OpenCL
+open Microsoft.FSharp.Quotations
+open GraphBLAS.FSharp.Backend
+open GraphBLAS.FSharp.Backend.Common
+open GraphBLAS.FSharp.Backend.Predefined
+
+module internal rec Expansion =
+    let run
+        (clContext: ClContext)
+        workGroupSize
+        (processor: MailboxProcessor<_>)
+        (matrixLeft: CSRMatrix<'a>)
+        (matrixRight: CSRMatrix<'a>)
+        (resultMatrix: CSRMatrix<'a>)
+        (times: Expr<'a -> 'a -> 'a>) =
+
+        let headFlags = getHeadFlags clContext workGroupSize processor matrixLeft
+
+        let rowPtrsForCols = getRowPtrsForCols clContext workGroupSize processor headFlags matrixLeft matrixRight
+        let positionsForResCols = getPositionsForResCols clContext workGroupSize processor headFlags matrixLeft matrixRight resultMatrix
+        processor.Post(Msg.CreateFreeMsg<_>(headFlags))
+
+        let resultHeadFlags = getResultHeadFlags clContext workGroupSize processor positionsForResCols matrixLeft.Columns.Length resultMatrix
+
+        let secondPositions = getSecondPositions clContext workGroupSize processor rowPtrsForCols positionsForResCols resultHeadFlags resultMatrix
+        processor.Post(Msg.CreateFreeMsg<_>(rowPtrsForCols))
+        processor.Post(Msg.CreateFreeMsg<_>(resultHeadFlags))
+        let firstPositions = getFirstPositions clContext workGroupSize processor positionsForResCols resultMatrix
+        processor.Post(Msg.CreateFreeMsg<_>(positionsForResCols))
+
+        let resultColumns = Gather.run clContext workGroupSize processor secondPositions matrixRight.Columns
+        let firstValues = Gather.run clContext workGroupSize processor firstPositions matrixLeft.Values
+        processor.Post(Msg.CreateFreeMsg<_>(firstPositions))
+        let secondValues = Gather.run clContext workGroupSize processor secondPositions matrixRight.Values
+        processor.Post(Msg.CreateFreeMsg<_>(secondPositions))
+
+        let resultValues = ClArray.zipWith clContext workGroupSize processor times firstValues secondValues
+        processor.Post(Msg.CreateFreeMsg<_>(firstValues))
+        processor.Post(Msg.CreateFreeMsg<_>(secondValues))
+
+        {
+            RowCount = resultMatrix.RowCount
+            ColumnCount = resultMatrix.ColumnCount
+            RowPointers = resultMatrix.RowPointers
+            Columns = resultColumns
+            Values = resultValues
+        }
+
+    let private getFirstPositions
+        (clContext: ClContext)
+        workGroupSize
+        (processor: MailboxProcessor<_>)
+        (positionsForResCols: ClArray<int>)
+        (resultMatrix: CSRMatrix<'a>) =
+
+        let positions = ClArray.create clContext workGroupSize processor resultMatrix.Columns.Length 0
+
+        updateFirstPositions clContext workGroupSize processor positionsForResCols positions
+
+        let max = <@ fun a b -> if a > b then a else b @>
```

Comment:
@dpanfilyonok Can we support ```max``` and ```min``` similar to [other math functions](https://github.com/YaccConstructor/Brahma.FSharp/blob/master/src/Brahma.FSharp.OpenCL.Translator/Body.fs#L134)? OpenCL has built-in max and min (https://www.khronos.org/files/opencl-1-2-quick-reference-card.pdf, Common Built-in Functions)